internal/pkg/cli/app_deploy.go

```diff
@@ -0,0 +1,423 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package cli
+
+import (
+	"errors"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/archer"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/aws/ecr"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/aws/session"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/docker"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/manifest"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/project"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/term/prompt"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/workspace"
+)
+
+// BuildAppDeployCommand builds the `app deploy` subcommand.
+func BuildAppDeployCommand() *cobra.Command {
+	input := &appDeployInput{
+		prompt: prompt.New(),
+	}
+
+	cmd := &cobra.Command{
+		Use:  "deploy",
+		Long: `Deploy an application to an environment.`,
+		Example: `
+  Deploy an application named "frontend" to a "test" environment.
+  $ archer app deploy --name frontend --env test`,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if err := input.init(); err != nil {
+				return err
+			}
+
+			if err := input.sourceInputs(); err != nil {
+				return err
+			}
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := input.deployApp(); err != nil {
+				return err
+			}
+
+			return nil
+		},
+		// PostRunE: func(cmd *cobra.Command, args []string) error {
+		// TODO: recommended actions?
+		// },
+	}
+
+	// TODO: make "project" a shared flag either at the `app` subcommand level or `root`?
+	cmd.Flags().StringVarP(&input.project, "project", "p", "", "project name")
+	cmd.Flags().StringVarP(&input.app, "app", "a", "", "application name")
+	cmd.Flags().StringVarP(&input.env, "env", "e", "", "environment name")
+	cmd.Flags().StringVarP(&input.imageTag, "tag", "t", "", "image tag")
+
+	return cmd
+}
+
+type appDeployInput struct {
```

Comment:
nit: What do you think about keeping the struct names ending with "Opts" for consistency?

Reply:
I don't prefer it since it isn't so much "options" as it is "user input", but I will change it for consistency.

Reply:
Gotcha, yeah I can see the naming weirdness. I believe "options" are used interchangeably with "flags" but our inputs can go beyond flags so I'd be down to change them in the future